File: code/py/src/modular_api/core/logger/logger.py
"""Structured JSON logger — RFC 5424 levels, request-scoped, zero dependencies.

Each log entry is a single JSON line with fixed fields:
``ts``, ``level``, ``severity``, ``msg``, ``service``, ``trace_id``.
Optional ``fields`` dict carries structured data from the caller.
"""

from __future__ import annotations

import json
import sys
import time
from enum import IntEnum
from typing import Callable, Protocol


class LogLevel(IntEnum):
    """RFC 5424 log levels in descending severity order.

    Filtering rule: if configured ``log_level = X``, only messages with
    ``value <= X`` are emitted.
    """

    emergency = 0  # system unusable
    alert = 1      # immediate action required
    critical = 2   # critical condition
    error = 3      # operation error, 5xx
    warning = 4    # abnormal condition, 4xx
    notice = 5     # normal but significant
    info = 6       # normal flow, 2xx/3xx
    debug = 7      # detailed diagnostics


class ModularLogger(Protocol):
    """Public logger interface exposed to UseCases.

    Each method corresponds to an RFC 5424 severity level.
    """

    @property
    def trace_id(self) -> str: ...
    def emergency(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def alert(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def critical(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def error(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def warning(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def notice(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def info(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...
    def debug(self, msg: str, *, fields: dict[str, object] | None = None) -> None: ...


# Signature for the output sink — default writes to stdout.
WriteFn = Callable[[str], object]


def _default_write(line: str) -> None:
    sys.stdout.write(line + "\n")


class RequestScopedLogger:
    """Per-request logger carrying ``trace_id`` with ``log_level`` filtering.

    Created by the logging middleware for each HTTP request and injected
    into the UseCase via the ``logger`` property.

    Pass a custom ``write_fn`` for testability (capture output without I/O).
    """

    def __init__(
        self,
        *,
        trace_id: str,
        log_level: LogLevel,
        service_name: str,
        write_fn: WriteFn = _default_write,
    ) -> None:
        self._trace_id = trace_id
        self._log_level = log_level
        self._service_name = service_name
        self._write_fn = write_fn

    @property
    def trace_id(self) -> str:
        return self._trace_id

    # ── Public API (8 RFC 5424 levels) ──────────────────────────────

    def emergency(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.emergency, msg, fields=fields)

    def alert(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.alert, msg, fields=fields)

    def critical(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.critical, msg, fields=fields)

    def error(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.error, msg, fields=fields)

    def warning(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.warning, msg, fields=fields)

    def notice(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.notice, msg, fields=fields)

    def info(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.info, msg, fields=fields)

    def debug(self, msg: str, *, fields: dict[str, object] | None = None) -> None:
        self._log(LogLevel.debug, msg, fields=fields)

    # ── Framework-internal: request/response logging ────────────────

    def log_request(self, *, method: str, route: str) -> None:
        """Emit a 'request received' log at info level."""
        self._log(LogLevel.info, "request received", extra={"method": method, "route": route})

    def log_response(
        self,
        *,
        method: str,
        route: str,
        status_code: int,
        duration_ms: float,
        extra: dict[str, object] | None = None,
    ) -> None:
        """Emit a 'request completed' log at the level determined by status_code."""
        self._log(
            _level_for_status(status_code),
            "request completed",
            extra={
                "method": method,
                "route": route,
                "status": status_code,
                "duration_ms": duration_ms,
                **(extra or {}),
            },
        )

    def log_unhandled_exception(self, *, route: str, duration_ms: float) -> None:
        """Emit an 'unhandled exception' log at error level. No stack trace by design."""
        self._log(LogLevel.error, "unhandled exception", extra={"route": route, "status": 500})

    # ── Internal ────────────────────────────────────────────────────

    def _log(
        self,
        level: LogLevel,
        msg: str,
        *,
        fields: dict[str, object] | None = None,
        extra: dict[str, object] | None = None,
    ) -> None:
        # Filtering: only emit if message level <= configured log_level.
        if level.value > self._log_level.value:
            return

        entry: dict[str, object] = {
            "ts": time.time(),
            "level": level.name,
            "severity": level.value,
            "msg": msg,
            "service": self._service_name,
            "trace_id": self._trace_id,
        }

        if extra is not None:
            entry.update(extra)
        if fields is not None:
            entry["fields"] = fields

        self._write_fn(json.dumps(entry))


def _level_for_status(status: int) -> LogLevel:
    """Map HTTP status code to RFC 5424 log level."""
    if status >= 500:
        return LogLevel.error
    if status >= 400:
        return LogLevel.warning
    if status >= 200:
        return LogLevel.info
    return LogLevel.notice  # 1xx

File: code/ts/modular_api/src/core/logger/logger.ts
// ============================================================
// core/logger/logger.ts
// LogLevel enum, ModularLogger interface, RequestScopedLogger.
// Mirror of logger.dart — RFC 5424, JSON to stdout, zero deps.
// ============================================================

/**
 * RFC 5424 log levels in descending severity order.
 *
 * Filtering rule: if configured `logLevel = X`, only messages with
 * `value <= X` are emitted. Higher values produce total silence.
 */
export enum LogLevel {
  emergency = 0, // system unusable
  alert = 1, // immediate action required
  critical = 2, // critical condition
  error = 3, // operation error, 5xx
  warning = 4, // abnormal condition, 4xx
  notice = 5, // normal but significant
  info = 6, // normal flow, 2xx/3xx
  debug = 7, // detailed diagnostics
}

/** Human-readable name for each LogLevel value. */
const LEVEL_NAME: Record<LogLevel, string> = {
  [LogLevel.emergency]: 'emergency',
  [LogLevel.alert]: 'alert',
  [LogLevel.critical]: 'critical',
  [LogLevel.error]: 'error',
  [LogLevel.warning]: 'warning',
  [LogLevel.notice]: 'notice',
  [LogLevel.info]: 'info',
  [LogLevel.debug]: 'debug',
};

/**
 * Public logger interface exposed to UseCases.
 *
 * Each method corresponds to an RFC 5424 severity level.
 * `fields` is an optional map of structured data attached to the log entry.
 */
export interface ModularLogger {
  /** Request-scoped trace ID for correlation. */
  readonly traceId: string;
  emergency(msg: string, fields?: Record<string, unknown>): void;
  alert(msg: string, fields?: Record<string, unknown>): void;
  critical(msg: string, fields?: Record<string, unknown>): void;
  error(msg: string, fields?: Record<string, unknown>): void;
  warning(msg: string, fields?: Record<string, unknown>): void;
  notice(msg: string, fields?: Record<string, unknown>): void;
  info(msg: string, fields?: Record<string, unknown>): void;
  debug(msg: string, fields?: Record<string, unknown>): void;
}

/** Function signature for the output sink — defaults to `console.log`. */
export type WriteFn = (line: string) => void;

/**
 * Per-request logger that carries `traceId` and respects `logLevel` filtering.
 *
 * Created by `loggingMiddleware` for each incoming HTTP request and injected
 * into the UseCase via the `logger` property.
 *
 * Accepts an optional `writeFn` for output — defaults to `console.log`.
 * In tests, pass a capturing function to inspect output without side-effects.
 */
/**
 * Builds platform-specific correlation fields from the ids the framework resolved.
 *
 * The framework emits open formats and nothing vendor-specific (roadmap invariant 7), so
 * a field like Google's `logging.googleapis.com/trace` — which needs a project id the
 * framework has no business knowing — is produced by the application.
 */
export type TraceFieldFormatter = (
  traceId: string,
  spanId: string | undefined,
) => Record<string, unknown>;

export class RequestScopedLogger implements ModularLogger {
  constructor(
    readonly traceId: string,
    readonly logLevel: LogLevel,
    readonly serviceName: string,
    private readonly writeFn: WriteFn = (line) => process.stdout.write(line + '\n'),
    /**
     * The caller's `X-Request-ID`, when it sent one. Preserved beside {@link traceId}
     * rather than promoted into it (runbook D6).
     */
    readonly requestId?: string,
    /**
     * The active server span's id, when tracing is configured.
     *
     * Known at construction because the tracing middleware is **outermost** (runbook D12
     * as reversed): the span already exists when `loggingMiddleware` creates this logger,
     * so nothing needs mutating afterwards.
     */
    readonly spanId?: string,
    /** Builds platform correlation fields. Undefined means none are emitted. */
    private readonly traceFieldFormatter?: TraceFieldFormatter,
  ) {}

  // ─── Public API (8 RFC 5424 levels) ──────────────────────────────

  emergency(msg: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.emergency, msg, { fields });
  }
  alert(msg: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.alert, msg, { fields });
  }
  critical(msg: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.critical, msg, { fields });
  }
  error(msg: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.error, msg, { fields });
  }
  warning(msg: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.warning, msg, { fields });
  }
  notice(msg: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.notice, msg, { fields });
  }
  info(msg: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.info, msg, { fields });
  }
  debug(msg: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.debug, msg, { fields });
  }

  // ─── Framework-internal: request/response logging ────────────────

  /** Emits a "request received" log at `info` level. */
  logRequest(opts: { method: string; route: string }): void {
    this.log(LogLevel.info, 'request received', {
      extra: { method: opts.method, route: opts.route },
    });
  }

  /** Emits a "request completed" log at the level determined by `statusCode`. */
  logResponse(opts: {
    method: string;
    route: string;
    statusCode: number;
    durationMs: number;
    extra?: Record<string, unknown>;
  }): void {
    this.log(RequestScopedLogger.levelForStatus(opts.statusCode), 'request completed', {
      extra: {
        method: opts.method,
        route: opts.route,
        status: opts.statusCode,
        duration_ms: opts.durationMs,
        ...(opts.extra ?? {}),
      },
    });
  }

  /** Emits an "unhandled exception" log at `error` level. No stack trace. */
  logUnhandledException(opts: { route: string }): void {
    this.log(LogLevel.error, 'unhandled exception', {
      extra: { route: opts.route, status: 500 },
    });
  }

  // ─── Internal ────────────────────────────────────────────────────

  private log(
    level: LogLevel,
    msg: string,
    opts: {
      fields?: Record<string, unknown>;
      extra?: Record<string, unknown>;
    } = {},
  ): void {
    // Filtering: only emit if the message level <= configured logLevel.
    if (level > this.logLevel) return;

    const entry: Record<string, unknown> = {
      ts: Date.now() / 1000,
      level: LEVEL_NAME[level],
      severity: level as number,
      msg,
      service: this.serviceName,
      trace_id: this.traceId,
      ...(this.spanId === undefined ? {} : { span_id: this.spanId }),
      ...(this.requestId === undefined ? {} : { request_id: this.requestId }),
    };

    // Only when there is trace context to point at. A platform field naming a trace
    // that does not exist is worse than no field.
    if (this.traceFieldFormatter !== undefined && this.spanId !== undefined) {
      Object.assign(entry, this.traceFieldFormatter(this.traceId, this.spanId));
    }

    if (opts.extra) Object.assign(entry, opts.extra);
    if (opts.fields) entry['fields'] = opts.fields;

    this.writeFn(JSON.stringify(entry));
  }

  /** Maps HTTP status code → RFC 5424 log level. */
  static levelForStatus(status: number): LogLevel {
    if (status >= 500) return LogLevel.error;
    if (status >= 400) return LogLevel.warning;
    if (status >= 200 && status < 400) return LogLevel.info;
    return LogLevel.notice; // 1xx
  }
}
